L12-csharp-interceptors.md

```diff
@@ -0,0 +1,212 @@
+
+Server-side interceptors derive from the new `ServerInterceptor` class
+implemented in `Grpc.Core.Interceptors` namespace.  Server interceptors are
+registered on individual service definitions as opposed to an entire server,
+though it might be sensible to provide a syntactic sugar for adding an
+interceptor chain to all services served by a `Server` instance.
+
+In particular, an instance of a class derived from `ServerInterceptor` is
+registered via an extension method for class `ServerServiceDefinition`.  This
+method is defined as an extension method to decouple the server interceptor
+machinery from the core gRPC infrastructure as much as possible.  This
+extension method wraps all of the existing handlers in the service definition
+with ones intercepted by the passed interceptor object.
+
+```csharp
+Server server = new Server
+{
+	Services = { Greeter.BindService(new GreeterImpl()).Intercept(new LogInterceptor()) },
+	Ports = { new ServerPort("localhost", Port, ServerCredentials.Insecure) }
+};
+server.Start();
+```
+
+Since support for this handler substitution was needed from the
+`ServerServiceDefinition` class itself, a general internal method
+`SubstituteHandlers` is added to actually apply the mapping given the wrapping
+logic as a delegate and return a new service definition object with the new
+handlers.  Beyond this, and slight modification to the definition of server
+call handlers, no special support from the core gRPC machinery was needed, and
+the fast path without any registered interceptor remains intact.
+
+Similar to client interceptors, server interceptors can override four functions
+`UnaryServerHandler`, `ClientStreamingServerHandler`,
+`ServerStreamingServerHandler`, and `DuplexStreamingServerHandler`.  Likewise,
+they all take a `next` argument that invokes the next interceptor in the chain
+or the actual RPC handler at the end of the chain:
+
+```csharp
+async Task<TResponse> UnaryServerHandler<TRequest, TResponse>(TRequest request, ServerCallContext context, UnaryServerMethod<TRequest, TResponse> next);
+async Task<TResponse> ClientStreamingServerHandler<TRequest, TResponse>(IAsyncStreamReader<TRequest> requestStream, ServerCallContext context, ClientStreamingServerMethod<TRequest, TResponse> next);
+async Task ServerStreamingServerHandler<TRequest, TResponse>(TRequest request, IServerStreamWriter<TResponse> responseStream, ServerCallContext context, ServerStreamingServerMethod<TRequest, TResponse> next);
+async Task DuplexStreamingServerHandler<TRequest, TResponse>(IAsyncStreamReader<TRequest> requestStream, IServerStreamWriter<TResponse> responseStream, ServerCallContext context, DuplexStreamingServerMethod<TRequest, TResponse> next);
+```
+
+In order to carry satellite data between interceptors and the handler, this
+proposal suggests adding an `Items` dictionary to `ServerCallContext` to hold
+arbitrary state.
+
+## Rationale
+
+The design goals where flexibility, decoupling and non-disruption to existing
```

Comment:
one design goal worth explicitly calling out is to not break the existing API surface.